mlflow/pyfunc/__init__.py

```diff
@@ -828,13 +828,189 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
+
+
+def _get_spark_primitive_types():
+    from pyspark.sql import types
+
+    return (
+        types.IntegerType,
+        types.LongType,
+        types.FloatType,
+        types.DoubleType,
+        types.StringType,
+        types.BooleanType,
+    )
+
+
+def _check_udf_return_struct_type(struct_type):
+    from pyspark.sql.types import StructType, ArrayType
+
+    primitive_types = _get_spark_primitive_types()
+
+    for field in struct_type.fields:
+        field_type = field.dataType
+
+        if isinstance(field_type, primitive_types):
+            pass
+        elif isinstance(field_type, ArrayType):
+            _check_udf_return_array_type(field_type, parent_struct_type=struct_type)
+        else:
+            raise MlflowException(
+                f"'spark_udf' return type does not support field type {field_type} "
+                "inside a struct type",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+
+def _check_udf_return_array_type(array_type, parent_struct_type):
+    from pyspark.sql.types import ArrayType, StructType
+
+    elem_type = array_type.elementType
+    primitive_types = _get_spark_primitive_types()
+
+    if isinstance(elem_type, primitive_types):
+        # 1D array
+        return
+
+    if isinstance(elem_type, StructType):
+        if parent_struct_type is None:
+            # Array of struct values.
+            _check_udf_return_struct_type(elem_type)
+            return
+
+        raise MlflowException(
+            "'spark_udf' return type does not support struct type nesting struct type.",
+            f"but the parent struct type {parent_struct_type} contains nested struct type "
+            f"{elem_type}.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+
+    if isinstance(elem_type, ArrayType) and isinstance(elem_type.elementType, primitive_types):
+        # 2D array
+        return
+
+    raise MlflowException(
+        f"'spark_udf' return type does not support the array type {array_type}, "
+        f"an array type only supports one dimensional array with primitive type {primitive_types} "
+        "elements or struct type elements, or double nested array consisting of "
+        "primitive type elements."
+    )
+
+
+def _check_udf_return_type(data_type):
+    from pyspark.sql.types import ArrayType, StructType
+
+    primitive_types = _get_spark_primitive_types()
+    if isinstance(data_type, primitive_types):
+        return
+
+    if isinstance(data_type, ArrayType):
+        _check_udf_return_array_type(data_type, parent_struct_type=None)
+        return
+
+    if isinstance(data_type, StructType):
+        _check_udf_return_struct_type(data_type)
+        return
+
+    raise MlflowException(
+        f"""Invalid data type: {data_type}, 'spark_udf' return type must be one of following types:
+Primitive types:
+ - int
+ - long
+ - float
+ - double
+ - string
+ - bool
+or compound types such as:
```

Comment:
We don't need `such as`. It gives me an impression that other types are also supported.